packages/dataverse/src/pointer.ts

```diff
@@ -56,18 +61,19 @@ export type PointerType<O> = {
  *
  */
 export type Pointer<O> = PointerType<O> &
-  (O extends UnindexableTypesForPointer
-    ? UnindexablePointer
-    : unknown extends O
-    ? UnindexablePointer
-    : O extends (infer T)[]
-    ? Pointer<T>[]
-    : O extends {}
-    ? {
-        [K in keyof O]-?: Pointer<O[K]>
-      } /*&
-        {[K in string | number]: Pointer<K extends keyof O ? O[K] : undefined>}*/
-    : UnindexablePointer)
+  PointerInner<Exclude<O, undefined>, undefined extends O ? undefined : never>
```

Comment:
Not sure how this works, could you explain in a comment?

Reply:
Hmm... I'll give it a whirl. First attempt to clarify: Essentially, the issue occurs is related to something like `Pointer<X | undefined>["..."]` resulting in an `UnindexablePointer`. This is because the first generic type is used to determine the overall kind of pointer (or list of pointers), but when that pointer is `| undefined` it gets off path. So, `Exclude<O, undefined>` removes any `undefined` variants in the union, and `undefined extends O ? undefined : never` results in `undefined` if `O` might be optional and `never` otherwise. So, now we can confidently use a non-undefined `O` in the ternary logic, while also ensuring that we keep track of whether it's optional via the `Optional` added param.

This is the part of the definition that matters (allowing us to remove optionality of each mapped key) without losing the information about the inner type being potentially undefined.
```ts
{
  [K in keyof O]-?: Pointer<O[K] | Optional>
}
```

Reply:
Applied a couple explanatory comments in the code